EIPS/eip-663.md

```diff
@@ -23,32 +23,39 @@ Furthermore, implementing higher level constructs, such as functions, on top of
+
+Introducing `SWAPN`, `DUPN` and `EXCHANGE` will provide an option to compilers to simplify accessing deep stack items.
 
 ## Specification
 
 We introduce two new instructions:
 
  1. `DUPN` (`0xe6`)
- 2. `SWAPN` (`0xe7`)
+ 1. `SWAPN` (`0xe7`)
+ 2. `EXCHANGE` (`0xe8`)
 
-If the code is legacy bytecode, both of these instructions result in an *exceptional halt*. (*Note: This means no change to behaviour.*)
+If the code is legacy bytecode, any of these instructions result in an *exceptional halt*. (*Note: This means no change to behaviour.*)
 
 If the code is valid EOF1, the following rules apply:
 
- 1. These instructions are followed by an 8-bit immediate value, which we call `imm`, and can have a value of 0 to 255. We introduce the variable `n` which equals to `imm + 1`.
+ 1. The instructions are followed by an 8-bit immediate value, which we call `imm`, and can have a value of 0 to 255.
+  1.1 In the case of `DUPN` and `SWAPN`, we introduce the variable `n` which equals to `imm + 1`.
+  1.2 In the case of `EXCHANGE`, we introduce the variable `n` which is equal to `imm >> 4 + 1`, and the variable `m` which is equal to `imm & 0x0F + 1` (i.e., the first and second nibbles of `imm`, converted to one-indexing).
 
- 2. Code validation is extended to check that no relative jump instruction (`RJUMP`/`RJUMPI`/`RJUMPV`) targets immmediate values of `DUPN` or `SWAPN`.
+ 2. Code validation is extended to check that no relative jump instruction (`RJUMP`/`RJUMPI`/`RJUMPV`) targets immmediate values of `DUPN`, `SWAPN` or `EXCHANGE`.
 
  3. The stack validation algorithm of [EIP-5450](./eip-5450.md) is extended:
-  3.1. Before `DUPN` if the current stack height is less than `n`, code is invalid. After `DUPN` stack height is incremented.
-  3.2. Before `SWAPN` if the current stack height is less than `n + 1`, code is invalid. After `SWAPN` stack height is not changed.
+  3.1. Before `DUPN` if the current stack height is less than `n`, code is invalid. After `DUPN`, the stack height is incremented.
+  3.2. Before `SWAPN` if the current stack height is less than `n`, code is invalid. After `SWAPN`, the stack height is unchanged.
+  3.2. Before `EXCHANGE` if the current stack height is less than `n + m`, code is invalid. After `EXCHANGE`, the stack height is unchanged.
 
  4. Execution rules:
   4.1. `DUPN`: the `n`'th stack item is duplicated at the top of the stack. (*Note: We use 1-based indexing here.*)
-  4.2  `SWAPN`: the `n + 1`th stack item is swapped with the top stack item.
+  4.2. `SWAPN`: the `n`'th stack item is swapped with the top of the stack. (*Note: We use 1-based indexing here.*)
+  4.3  `EXCHANGE`: the `n`'th stack item is swapped with the `n + m`'th stack item. (*Note: We use 1-based indexing here.*)
```

Comment:
This limits the swapping of `nth` item to only items `[n+1, n+16]`, so it can go 32 items deep but only with max distance 16 between items. E.g. you can't swap 1st with 32d.

Is this acceptable limitation for you?

Reply:
yes, we might want to change the encoding still.